presto-main/src/main/java/com/facebook/presto/execution/SqlQueryExecution.java

```diff
@@ -437,12 +456,15 @@ public void fail(Throwable cause)
     }
 
     @Override
-    public Duration waitForStateChange(QueryState currentState, Duration maxWait)
-            throws InterruptedException
+    public ListenableFuture<QueryOutputInfo> getOutputInfo()
     {
-        try (SetThreadName ignored = new SetThreadName("Query-%s", stateMachine.getQueryId())) {
-            return stateMachine.waitForStateChange(currentState, maxWait);
-        }
+        return Futures.nonCancellationPropagating(outputInfo);
```

Comment:
You were already thinking about cancellation so you thought whether you want to expose cancellation to the caller or not. In general -- this is _very_ not obvious and it seems all too easy to return thing like `outputInfo` directly.